### ISSUE TITLE
Numeric alias in Symfony routing

### ISSUE THREAD

**Issue by aschempp** [contao/contao#2306]
> I've just come across this issue in the wild and it sent me down a bit of a rabbit hole.
> 
> 4.4.50 and 4.9.5 are also affected by the bug described in the initial message.
>
> I've tried to follow the decisions in this thread, but I'm confused as to why the fix suggested by @leofeyer in the initial message ("trigger an error if someone enters a numeric alias") was not adopted until Contao can work with numeric aliases. Is it because the issue was never technically confirmed for an LTS?
> 
> Side note: Interestingly, an alias like `1234e0` also causes the problem because it `is_numeric`. The way Contao checks for IDs vs aliases was fixed in d3839c398db358d029ea8ca0f3ec7308d25cf043, but the Symfony CMF Router integration that was developed in parallel was not adjusted accordingly, and so the check there is still `is_numeric`, e. g.:  https://github.com/contao/contao/blob/f1530fc6b67f8a2d104943d9d0397a8a23dd5616/core-bundle/src/Routing/RouteProvider.php#L475

_Originally posted by @gmpf in https://github.com/contao/contao/issues/1015#issuecomment-692046902_

**Comment by aschempp** [comment]
Good catch! You mean we should use `preg_match('/^[1-9]\d*$/', $varId)` in the Symfony router as well?

However, you said this also affects Contao 4.4, and that version does not use CMF Routing?

**Comment by gmpf** [comment]
> Good catch! You mean we should use `preg_match('/^[1-9]\d*$/', $varId)` in the Symfony router as well?

I do, because it looks like that was the intent of making that change as a part of d3839c398db358d029ea8ca0f3ec7308d25cf043. @ausi would likely have a definitive answer on that.

> However, you said this also affects Contao 4.4, and that version does not use CMF Routing?

It kind of depends on what you mean by "this", but it also doesn't matter because both versions are affected by both things. :D 

The main issue in #1015 was that you could give a page an ID-like alias, which doesn't work (page doesn't get found). That's what I meant with "4.4.50 and 4.9.5 are also affected"; page aliases that look like IDs didn't work before the introduction of CMF Routing and they also don't work now.

The other thing, which is everything after "Side note:", is about using `is_numeric` vs. that regex to determine whether something is an ID or an alias. Changing from `is_numeric` to the regex fixes false positives for things that are numeric, but can't be IDs, like e notation (`1234e0`). 4.4 is affected by this because 4.4 still uses `is_numeric` (the fix from d3839c398db358d029ea8ca0f3ec7308d25cf043 never got backported). This doesn't fix the main issue in #1015 though.

**Comment by ausi** [comment]
> > Good catch! You mean we should use `preg_match('/^[1-9]\d*$/', $varId)` in the Symfony router as well?
> 
> I do, because it looks like that was the intent of making that change as a part of [d3839c3](https://github.com/contao/contao/commit/d3839c398db358d029ea8ca0f3ec7308d25cf043). @ausi would likely have a definitive answer on that.

I agree, aliases that match `/^[1-9]\d*$/` are considered page IDs in Contao, everything else is considered to be an alias string.

> The main issue in #1015 was that you could give a page an ID-like alias, which doesn't work

I think this should be fixed by prepending the following code to https://github.com/contao/contao/blob/5aae3bff5772702db67a8d6588c39629da419845/core-bundle/src/Resources/contao/dca/tl_page.php#L1153

```php
elseif (preg_match('/^[1-9]\d*$/', $varValue)) {
	throw new Exception(sprintf($GLOBALS['TL_LANG']['ERR']['aliasNumeric'], $varValue));
}
```
